spec/formatting.md

```diff
@@ -18,3 +18,248 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+Errors in messages and their formatting may occur and be detected
+at multiple different stages of their processing.
+Where available,
+the use of validation tools is recommended,
+as early detection of errors makes their correction easier.
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided into the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is not well-formed.
+
+  Example invalid messages resulting in a Syntax error:
+
+  ```
+  {Missing end brace
+  ```
+
+  ```
+  {Unknown {#placeholder#}}
+  ```
+
+  ```
+  let $var = {(no message body)}
+  ```
+
+- **Data Model errors** occur when a message is invalid due to
+  violating one of the semantic requirements on its structure:
+
+  - **Variant Key Mismatch errors** occur when the number of keys on a Variant
+    does not equal the number of Selectors.
+
+    Example invalid messages resulting in a Variant Key Mismatch error:
+
+    ```
+    match {$one}
+    when 1 2 {Too many}
```

Comment:
Having too few or too many keys (compared to the match expression) is clearly a syntax error, and should be in that category. It's like calling a function in C or Java with too few or too many parameters; not a semantic or data model error, but rather one that can be detected purely because of malformed syntax.

Reply:
The intent in differentiating this as a "data model" rather than "syntax" error is that it allows us to apply the same error check independently of the syntax used by the message. Here's what I said about this earlier in https://github.com/unicode-org/message-format-wg/pull/320#discussion_r1068067671:

> Let's say that there's a FooFormat9000 that's exactly like MF1, except that it doesn't require an "other" case for its selectors and defaults to an empty string instead. Now, if a FooFormat9000 message is parsed into an MF2 data model for use with an MF2 runtime, the code doing so might not add the expected default variant to the data model. When the MF2 runtime discovers this, it needs to emit a [Data Model] Error rather than a Syntax Error, because it doesn't know anything about FooFormat9000 where this syntax is valid.

An MF2 syntax parser would be expected to emit the data model errors during its work, as this is included in the spec (line 176):
> Syntax and Data Model errors must be emitted as soon as possible.

This separation also matches the differentiation between "well-formed" and "valid" messages in the [syntax spec](https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md#productions).